source/unified-test-format/unified-test-format.rst

```diff
@@ -442,7 +475,14 @@ The structure of this object is as follows:
     connecting to multiple mongos servers is necessary to test session
     pinning.
 
-    This option has no effect for non-sharded topologies.
+    If this option is true or unset and the topology type is
+    ``LoadBalanced``, the test runner MUST connect the MongoClient to a load
+    balancer that is balancing multiple servers. If it is false and the
+    topology type is ``LoadBalanced``, the test runner MUST connect the
+    MongoClient to a load balancer that is only fronting a single server.
+
+    This option has no effect for topologies that are not sharded or load
+    balanced.
```

Comment:
I think I missed this addition in an earlier review. Perhaps it explains all of the `useMultipleMongoses: true` lines that I asked about in the test files above.

It seems odd to expand the use of `useMultipleMongoses` in this way, as it was previously only used to test transactions and session pinning (unless I'm mistaken). Reading #939, I see load balancers currently front:

> Any MongoDB service that can run behind a load balancer, including but not limited to mongos, Atlas Proxy, Elastic Front End.

How feasible is it for drivers to detect what (or how many) servers are behind the load balancer?

And with regard to existing use of this field for transaction tests, do you anticipate a conflict if we start running those tests against a load balancer? Perhaps it's not a concern if none of those tests include `load-balancer` in their accepted topologies?

I wonder if it'd be safer to introduce a separate option for this (e.g. `loadBalanceMultipleServers`), which is only relevant if the topology is load-balanced (similar to how `useMultipleMongoses` is currently ignored for non-sharded topologies).

Reply:
> How feasible is it for drivers to detect what (or how many) servers are behind the load balancer?

AFAIK, neither of these are possible or desirable as we don't want to special case any mongos-like services and mongos servers are not aware of each other.

> do you anticipate a conflict if we start running those tests against a load balancer

If we find that any of the existing transactions tests fail against a load balancer, we should explicitly mark them as only being valid for replica sets and/or sharded clusters via the `topologies` requirement. My guess is many of these tests will already be skipped against LB clusters because they include topology-specific runOnRequirement blocks to account for the existence of `failCommand` on different server versions.

> I wonder if it'd be safer to introduce a separate option for this (e.g. loadBalanceMultipleServers)

I briefly discussed this with @jyemin when designing the unified test format changes for this spec. We opted to reuse `useMultipleMongoses` because the spirit of the constraint is the same: we want the test to be communicating with more than one mongos server to test some special behavior. There is a concern that it's not possible to verify that there are actually multiple mongos's behind the LB during testing, so we have to rely on orchestration scripts to ensure that. I'm not sure how adding a new option that's basically the same thing but only applies to LB clusters would help.

Reply:
Points noted on the first two replies.

> We opted to reuse useMultipleMongoses because the spirit of the constraint is the same: we want the test to be communicating with more than one mongos server to test some special behavior. There is a concern that it's not possible to verify that there are actually multiple mongos's behind the LB during testing, so we have to rely on orchestration scripts to ensure that. I'm not sure how adding a new option that's basically the same thing but only applies to LB clusters would help.

I'm reading that you want to test communication with "more than one mongos" and are concerned about not being able to verify the number of mongos nodes behind a LB. That suggests to me that `useMultipleMongoses` is still intended to apply to mongos specifically.

I originally misread this change as amending `useMultipleMongoses` to also cover "multiple load balancers", but that seems not to be the case. Re-reading the text above, I realize it isn't actually saying anything about the number of load balancers -- just the number of servers behind them. Furthermore, I gather it's assuming that the type of server will always be mongos and disregard other possible types noted in #939 (e.g. elastic).

Here is a very direct series of questions to help me understand more clearly:

 * With the suggested changes above, does `useMultipleMongoses` remain _only_ concerned with whether or not the driver is communicating with multiple mongos nodes (either directly or through one or more load balancers)? If "yes", why is the presence and/or number of load balancers irrelevant?
 * `useMultipleMongoses` has historically been used to test session pinning. I see a few references to "pin" in #939 but it's not clear to me how much it differs from what is currently dictated by [Transactions: Mongos Pinning](https://github.com/mongodb/specifications/blob/master/source/transactions/transactions.rst#mongos-pinning). Does pinning apply to drivers that connect to multiple load balancers? If so, is the type of server(s) behind the load balancers irrelevant to this pinning?

If there is going to be an assumption that any load balancer topology is fronting multiple mongoses (such that a load balancer topology is sufficient to satisfy `useMultipleMongoses: true` with no further introspection), I'd argue that the right course of action would be to add `load-balanced` as a permitted `runOnRequirement` for those existing tests that currently require a sharded cluster and specify `useMultipleMongoses: true`. AFAIK, the `runOnRequirement` would need to be changed anyway to permit those tests to run against an LB (see: [transactions/tests/mongos-pin-auto.yml](https://github.com/mongodb/specifications/blob/master/source/transactions/tests/mongos-pin-auto.yml)). Furthermore, this would entail no changes to how the test runner evaluates `useMultipleMongoses` (it would continue to be ignored for non-sharded toplogies).

Reply:
I pushed some changes to describe passing in multiple URIs to the test runner in the "Initializing a Test Runner" section and modified the text for ``useMultipleMongoses`` to match. I skipped the bit about defining "mongos" at the top of the spec because we're still only talking about testing against "real" mongos servers right now. The serverless testing project will be the first to test against mongos-like servers so I'd like to defer this change until that project is done.